### PR TITLE
Add detection of broken Javadoc paths to the FindBrokenInternalLinks task

### LINKED ISSUE #22836: Dead Javadoc links in Gradle user guide
Two Javadoc references from the [user guide](https://docs.gradle.org/7.6-rc-2/userguide/upgrading_version_7.html#updates_to_attribute_disambiguation_rules_related_methods), both in versions 7.6 RC 2 and 3, return `404 Not found`.

Screenshot 2022-11-23 at 17 02 28

There might be other cases, so it should be good to automate testing for dead links.

### PR DESCRIPTION
**Pull request by tresat**
This will avoid issues where invalid javadoc paths create links with 404s in published documentation.

This change found several other bad javadoc links.  Links with bad paths have been adjusted, absent links changed to `code.block.text` with a (deprecated) or (removed) note.

Fixes #22836

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test BD

**Comment by tresat** [comment]
@bot-gradle test BD

**Comment by tresat** [comment]
@bot-gradle test BD

**Review by nathan-contino — APPROVED** [review]
LGTM! Thanks, this is rad!

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc [hunk 1]
```diff
@@ -1223,7 +1223,7 @@ Existing clients should migrate to the <<publishing_maven.adoc#publishing_maven,
 The concept of conventions is outdated and superseded by extensions. To reflect this in the Gradle API, the following elements are now deprecated:
 
 - link:{javadocPath}/org/gradle/api/Project.html#getConvention--[Project.html#getConvention()]
-- link:{javadocPath}/org/gradle/api/internal/HasConvention.html[HasConvention]
+- `org.gradle.api.internal.HasConvention` (deprecated)
```
Comment (jvandort) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc:1226]: How come we've removed this link? Do we not generate javadoc for deprecated types?

File: build-logic/documentation/src/main/groovy/gradlebuild/docs/FindBrokenInternalLinks.java [hunk 2]
```diff
@@ -151,6 +128,55 @@ private void hasDeadLink(File baseDir, File sourceFile, Map<File, List<Error>> e
...
+    private void gatherDeadJavadocLinksInLine(File sourceFile, String line, int lineNumber, List<Error> errorsForFile) {
+        Matcher matcher = javadocLinkPattern.matcher(line);
+        while (matcher.find()) {
+            MatchResult linkMatcher = matcher.toMatchResult();
+            String link = linkMatcher.group(1);
+            File referencedFile = new File(getJavadocRoot().get().getAsFile(), link);
+            if (!referencedFile.exists() || referencedFile.isDirectory()) {
+                String errMsg = "Missing Javadoc file for " + link + " in " + sourceFile.getName();
+                if (link.startsWith("javadoc")) {
+                    errMsg += " (Did you mean to include the leading `javadoc` path component?)";
```
Comment (jvandort) [inline build-logic/documentation/src/main/groovy/gradlebuild/docs/FindBrokenInternalLinks.java:172]: This check seems pretty arbitrary now. I get that we had existing examples which included the `javadoc` component, but now that we have this check, how often are we going to run into this case? 

Though, if we leave this in, I think the message wording is confusing. It almost feels like its suggesting to add the path component. Consider updating the text to suggest removing the component. `Did you mean to include` sounds like its suggesting that I include the component in the URL.

**Review by jvandort — APPROVED** [review]
LGTM other than some minor comments

File: build-logic/documentation/src/main/groovy/gradlebuild/docs/FindBrokenInternalLinks.java (hunk 2 above)
Comment (tresat) [reply build-logic/documentation/src/main/groovy/gradlebuild/docs/FindBrokenInternalLinks.java:172]: This check is primarily for me. I've done this like 5 times. 🤭

Good point on the message, I'll update.

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc (hunk 1 above)
Comment (tresat) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc:1226]: We do not, it's a dead link.

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**